### ISSUE TITLE
Infix printing for `#=` and `##`

### ISSUE THREAD

**Issue by chenglou** [facebook/reason#650]
cc @bobzhang 

The BuckleScript readme example prints in prefix notation for reason:

```rust
let port = 3000;

let hostname = "127.0.0.1";

let create_server http => {
  let server = (## http createServer) (
  (
    fun req resp => {
      #= (## resp statusCode) 200;
      (## resp setHeader) "Content-Type" "text/plain";
      (## resp _end) "Hello world\n"
    }
  )
  [@bs]
  );
  (## server listen)
    port
    hostname
    (
    (
      fun () => Js.log (
        "Server running at http://" ^ hostname ^ ":" ^ Pervasives.string_of_int port ^ "/"
      )
    )
    [@bs]
    )
};

let () = create_server Http_types.http;
```

**Comment by jordwalke** [comment]
Oh ## is a standard function identifier?

**Comment by bobzhang** [comment]
OCaml treats `#_` operator in a special way, let me find the link

**Comment by bobzhang** [comment]
http://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec253
Operator names starting with a # character and containing more than one # character are reserved for extensions.

**Comment by jordwalke** [comment]
Thanks for the info.

**Comment by chenglou** [comment]
@yunxing any solution for this? I'm not familiar with this at all

**Comment by Schmavery** [comment]
Do we maybe just need to recognize it as an identifier and make sure reason doesn't die on it?  Then we can run the regular ## preprocessor on the .ml code?
This seems to be blocking the regular use of js bindings.  When we wrote reason code with js_of_ocaml we had to use the workaround of writing the code that the preprocessor generates (which works fine but ## is more convenient).

**Comment by jordwalke** [comment]
It should be as easy as considering this of `Infix` precedence (there's a function in the printer that checks to see if the leading character is infix - we'd just update it to also look for `#`). This is like a GoodFirstTask.

**Comment by yunxing** [comment]
@chenglou Sorry just saw this. 

@bobzhang @Schmavery I guess I'm a bit confused here. From the parser it seems like `#_` is still parsed as infix operators? How does this related to extensions ?

 https://github.com/ocaml/ocaml/blob/d6c72fc82a8950851da5f121b80a4e1c8bff377a/parsing/parser.mly#L1569

**Comment by bobzhang** [comment]
@Drup may know more about it

**Comment by Drup** [comment]
`#...` are infix operators (with priority higher than function application, just like `#`).
`#...#...` (with at least 2 `#`) are infix operators that are reserverd for ppx extensions. It's not possible to define them.

```ocaml
# let (#=#=) x y = x + y ;;
Error: '#=#=' is not a valid value identifier.  
```

**Comment by jordwalke** [comment]
Thanks for the info Drup. Though they are reserved, they seem to be represented as standard infix application. The check that they are used only by ppx seems to be deferred to a later stage of compilation. Makes sense.

**Comment by yunxing** [comment]
@Drup Makes sense now. Thank you!

**Comment by yunxing** [comment]
Closing this as it should be fixed by #657.
